File: Leetcode-5 Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        s2=s[::-1]
        self.req=[0,0]
        def ss(i,n):
            if n%2==0:
                self.req=[i-n//2+1,i+n//2+1]
            else:
                self.req=[i-n//2,i+n//2+1]
        l1=len(s)+1
        ll=[[0 for i in range(l1)] for j in range(l1)]
        out=0

        for i in range(1,l1):
            for j in range(1,l1-i+1):
                if j==l1-i:
                    ll[i][j]=ll[i-1][j-1]+1
                elif s2[i-1]==s[j-1]:
                    ll[i][j]=ll[i-1][j-1]+2

                if ll[i][j]>out and (i+j==l1 or i+j+1==l1):
                    out=ll[i][j]
                    ss(j-1,ll[i][j])
            #print(ll[i])

        return s[self.req[0]:self.req[1]]

File: Leetcode-5 Longest Palindromic Substring.java
class Solution {
    public void ss(int i,int n,int[] req){
            if (n%2==0)
                req[0]=i-n/2+1;
            else
                req[0]=i-n/2;
            
            req[1]=i+n/2+1;
        }
    public String longestPalindrome(String s) {
        int req[]=new int[]{0,0};
        
        int out=0;
        int l1=s.length()+1;
        int ll[][]= new int[l1][l1];
        for (int i=l1-1;i>0;i--){
            for(int j=1;j<=i;j++){
                if (i==j)
                    ll[l1-i][j]=ll[l1-i-1][j-1]+1;
                else if (s.charAt(j-1)==s.charAt(i-1))
                    ll[l1-i][j]=ll[l1-i-1][j-1]+2;
                
                if (ll[l1-i][j]>out && (i==j+1 || i==j)){
                    out=ll[l1-i][j];
                    ss(j-1,ll[l1-i][j],req);
                }
            //System.out.print(ll[i][j]);
            }
        //System.out.println();
        }
        
        return s.substring(req[0],req[1]);
    }
}
